File: app/src/test/java/com/localphoto360/app/util/EquirectangularMathTest.java
package com.localphoto360.app.util;

import org.junit.Assert;
import org.junit.Test;

public class EquirectangularMathTest {
    @Test
    public void lookNorthMapsToImageCenter() {
        float[] uv = EquirectangularMath.lookToUv(0f, 0f, 1f);
        Assert.assertEquals(0.5f, uv[0], 0.001f);
        Assert.assertEquals(0.5f, uv[1], 0.001f);
    }

    @Test
    public void lookEastMapsToThreeQuarters() {
        float[] uv = EquirectangularMath.lookToUv(1f, 0f, 0f);
        Assert.assertEquals(0.75f, uv[0], 0.001f);
    }

    @Test
    public void zenithAndNadir() {
        float[] zenith = EquirectangularMath.lookToUv(0f, 1f, 0f);
        float[] nadir = EquirectangularMath.lookToUv(0f, -1f, 0f);
        Assert.assertEquals(0.0f, zenith[1], 0.01f);
        Assert.assertEquals(1.0f, nadir[1], 0.01f);
    }

    @Test
    public void uvRoundTrip() {
        float[] look = EquirectangularMath.uvToLook(0.25f, 0.5f);
        float[] uv = EquirectangularMath.lookToUv(look[0], look[1], look[2]);
        Assert.assertEquals(0.25f, uv[0], 0.002f);
        Assert.assertEquals(0.5f, uv[1], 0.002f);
    }

    @Test
    public void oppositeLooksArePiApart() {
        float distance = EquirectangularMath.angularDistanceRad(0f, 0f, (float) Math.PI, 0f);
        Assert.assertTrue(Math.abs(distance - (float) Math.PI) < 0.01f);
    }

    @Test
    public void rotationMatrixMatchesYawPitch() {
        float yaw = (float) Math.toRadians(90.0);
        float pitch = (float) Math.toRadians(20.0);
        float[] matrix = EquirectangularMath.rotationMatrixFromYawPitch(yaw, pitch);
        Assert.assertEquals(yaw, Math.atan2(matrix[2], matrix[8]), 0.02);
        Assert.assertEquals(pitch, Math.asin(EquirectangularMath.clamp(matrix[5], -1f, 1f)), 0.02);
    }
}

File: app/src/test/java/com/localphoto360/app/util/EquirectangularMathTest.kt
package com.localphoto360.app.util

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.math.abs

class EquirectangularMathTest {

    @Test
    fun lookNorthMapsToImageCenter() {
        val (u, v) = EquirectangularMath.lookToUv(0f, 0f, 1f)
        assertEquals(0.5f, u, 0.001f)
        assertEquals(0.5f, v, 0.001f)
    }

    @Test
    fun lookEastMapsToThreeQuarters() {
        val (u, _) = EquirectangularMath.lookToUv(1f, 0f, 0f)
        assertEquals(0.75f, u, 0.001f)
    }

    @Test
    fun zenithAndNadir() {
        val zenith = EquirectangularMath.lookToUv(0f, 1f, 0f)
        val nadir = EquirectangularMath.lookToUv(0f, -1f, 0f)
        assertEquals(0.0f, zenith.second, 0.01f)
        assertEquals(1.0f, nadir.second, 0.01f)
    }

    @Test
    fun uvRoundTrip() {
        val look = EquirectangularMath.uvToLook(0.25f, 0.5f)
        val (u, v) = EquirectangularMath.lookToUv(look[0], look[1], look[2])
        assertEquals(0.25f, u, 0.002f)
        assertEquals(0.5f, v, 0.002f)
    }

    @Test
    fun oppositeLooksArePiApart() {
        val distance = EquirectangularMath.angularDistanceRad(0f, 0f, Math.PI.toFloat(), 0f)
        assertTrue(abs(distance - Math.PI.toFloat()) < 0.01f)
    }
}
